CUDNN v8 Implementation of Convolution Kernels

Tom-Zheng opened the pull request:
<!-- Demo: https://github.com/PaddlePaddle/Paddle/pull/24810 -->
### PR types
<!-- One of [ New features | Bug fixes | Function optimization | Performance optimization | Breaking changes | Others ] -->
New features
### PR changes
<!-- One of [ OPs | APIs | Docs | Others ] -->
OPs
### Describe
<!-- Describe what this PR does -->
We are making a transition from the legacy CUDNNv7 APIs into the latest CUDNN frontend API which is recommended for CUDNNv8 and later. CUDNN frontend API provides easier programming interface along with modern functionalities like autotuning and errata filter(blocks certain engine configs by json files), providing much more flexibility.
As a first step, this merge implements CUDNNv8 APIs for the convolution operator, both forward and backward.

- To build paddle with CUDNNv8 APIs, set the build option `WITH_CUDNN_FRONTEND` to `ON`. Currently, the default value of this option is `OFF`.
- We still use legacy APIs by default. To enable CUDNNv8 APIs at runtime, set the environment variable `FLAGS_enable_cudnn_frontend=1`.

Tom-Zheng committed:
Refactor conv_kernel and conv_grad_kernel to provide interface for
CUDNNv8 implementation

Tom-Zheng committed:
Fix macro

Tom-Zheng committed:
Add implementation for conv_kernel and conv_grad_kernel

Tom-Zheng committed:
Modification after rebase onto latest develop

Tom-Zheng committed:
Modify plan cache to comply with the API of phi::autotune

Tom-Zheng committed:
Fix build errors

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,140 @@
+#include "paddle/fluid/platform/flags.h"
+#include "paddle/phi/backends/dynload/cudnn_frontend.h"
+
+DECLARE_int64(cudnn_exhaustive_search_times);
```
`autotune`功能下，不需要使用`cudnn_exhaustive_search_times`控制搜索次数。我们通过设置自动调优区间来控制搜索次数，默认调优区间为`[1, 10)`，在调优区间内会开启穷举搜索。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,140 @@
+namespace phi {
+namespace autotune {
+
+class CudnnFrontendPlanCache {
```
当前版本，`AlgorithmsCache`和`CudnnAlgorithmsCacheMap`存在太多的共同代码，https://github.com/PaddlePaddle/Paddle/pull/47667 尝试通过继承的方式重写，后续可更新下。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,140 @@
+  }
+
+ private:
+  static cudnn_frontend::feature_vector_t MakeKey(
```
cudnn frontend API本身就提供了cache key的计算方式？相比当前`ConvCacheKey`，主要差别在哪里呢？

https://github.com/PaddlePaddle/Paddle/blob/fa874a463f93e275b46faf3c345a75d36c214290/paddle/phi/kernels/autotune/cache.h#L76

Tom-Zheng replied:
本质上没有大的区别，只是API的区别。cudnn v7的编程模式是自己去维护convolution 相关的descriptors, 自己去做数据结构去存各种参数，自己去实现key和cache。CUDNN Frontend API提供了更多的封装，面向对象的设计。在v7里面，这个ConvCacheKey 是通过ConvArgs里面的ConvertToConvCacheKey实现的。但在v8里面我们不再用这个ConvArgs管理descriptor了，而是用v8的class。那么最方便的就是使用提供好的key实现方式。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,140 @@
+        std::make_pair(MakeKey(op_graph, use_addto), plan.GetEngineConfig()));
+  }
+
+  bool IsStable(const cudnn_frontend::OperationGraph& op_graph,
```
没太理解，这个函数的作用是什么？

Tom-Zheng replied:
有可能穷举搜索的结果不稳定，例如某5次搜索得到的algo是： [1,1,0,0,0]，相比于只信任某一次搜索的结果，我们提供一个选项，可以设定一个saturation count，设为N. 某一个algo至少在N次搜索中得到最佳结果，才会被加到cache中去。上面的例子，如果N取3，那么algo1就不会加入cache，0会被认为是最快的algo并加入cache. 这个函数是用来判断是否达到saturation count的，如果返回true才会加到cache。

Xreki commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,258 @@
+    return out_array;
+  }
+
+  static void GenerateStrides(const int64_t* dimA,
```
Paddle中的代码，请尽量符合Google编码规范，变量名使用`xxx_xxx`这种小写的驼峰式命名。另外，`dimA`参数类型能否使用`std::vector<int64_t>`？`strides`能否作为返回值？

Xreki commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,258 @@
+      const phi::DenseTensor* tensor,
+      int64_t id,
+      cudnnTensorFormat_t layout_format) {
+    auto dims = phi::vectorize<int>(tensor->dims());
```
为了调用`TransformDimOrder`，这里强制把dims转成`std::vector<int>`，感觉没有必要，这里可以直接用`phi::vectorize<int64_t>`，同时可以给`TransformDimOrder`加上模板，设置默认类型为`int`。

Xreki commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,258 @@
+      const std::vector<int>& dilation) {
+    uint64_t conv_dim = stride.size();
+    cudnnDataType_t compute_type =
+        (dataType == CUDNN_DATA_DOUBLE) ? CUDNN_DATA_DOUBLE : CUDNN_DATA_FLOAT;
```
v7接口，使用的计算类型和这里一样吗？

Tom-Zheng replied:
一样： 
https://github.com/PaddlePaddle/Paddle/blob/e12a905e9999b87315f27ca99ade38f5fdf22a74/paddle/fluid/platform/device/gpu/cuda/cudnn_desc.h#L271

Xreki commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,258 @@
+#include "paddle/phi/backends/gpu/gpu_context.h"
+#include "paddle/phi/core/dense_tensor.h"
+#include "paddle/phi/kernels/autotune/switch_autotune.h"
+#include "paddle/phi/kernels/gpudnn/conv_cudnn_v7.h"
```
若只是为了调用`CalcWorkspaceLimitInBytes`而引入`conv_cudnn_v7.h`头文件，可以把几个函数挪到`conv_gpudnn_base.h`中。

Xreki replied:
该include可以移除了。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel.cu:
```diff
@@ -592,6 +599,63 @@ void ConvCudnnGradKernel(const Context& ctx,
+                                 groups,
+                                 data_format,
+                                 input_grad,
+                                 filter_grad);
```
可以在`if`里面提前`return`，避免`ConvCudnnGradKernelImplV7`调用的代码重复写2次。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,552 @@
+    }
+  }
+
+  PADDLE_ENFORCE_EQ(
```
可以用`PADDLE_THROW`

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,552 @@
+        padding_common[i] = paddings[2 * i];
+      }
+    }
+  }
```
此处大断代码和v7分支重复，会使后续功能维护变得很困难。请考虑一下细粒度一些的封装方案。

Tom-Zheng replied:
已重构。

Tom-Zheng committed:
Review changes

Tom-Zheng committed:
Refactor to reduce duplicate code

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,138 @@
+#include <string>
+#include <vector>
+
+#include "paddle/fluid/platform/flags.h"
```
不需要include这个头文件。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,138 @@
+                bool use_addto = false) {
+    bool ret = false;
+    std::lock_guard<std::mutex> lock(*cache_mutex_);
+    auto key = op_graph.getFeatureVector();
```
key后面没有被用到吧，可以删除。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,138 @@
+    bool ret = false;
+    std::lock_guard<std::mutex> lock(*cache_mutex_);
+    auto key = op_graph.getFeatureVector();
+    if (map_.count(MakeKey(op_graph, use_addto)) > 0) {
```
`MakeKey`这个函数需要多次重复调用吗，会有开销吗？

Tom-Zheng replied:
需要多次调用。该函数定义是implicitly inline的，所以还好。

Xreki commented on paddle/phi/kernels/autotune/cache_cudnn_frontend.h:
```diff
@@ -0,0 +1,138 @@
+    auto plan = cudnn_frontend::ExecutionPlanBuilder()
+                    .setHandle(handle)
+                    .setEngineConfig(engine_config, op_graph.getTag())
+                    .build();
```
Cache里面存的并不是`Plan`对象，而是`Plan`的`Config`？使用Config构建Plan的过程，是不是放在外面，即`CudnnFrontendConvHelper`里面比较好？

Tom-Zheng replied:
已移动。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,303 @@
+#include "paddle/fluid/framework/eigen.h"
+#include "paddle/fluid/platform/cudnn_workspace_helper.h"
+#include "paddle/fluid/platform/float16.h"
+#include "paddle/fluid/platform/profiler.h"
```
尽量不要inlcude fluid目录下面的头文件吧，这4个头文件应该都可以直接移除？建议再精简下include的头文件。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,303 @@
+
+template <typename T>
+void CudnnConvBwdDataV8(DenseTensor* dy_tensor,
+                        DenseTensor* w_tensor,
```
输入Tensor、且不允许修改的，加一下const修饰。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,303 @@
+        plan_cache_bwd_data.InsertPlan(op_graph, plan, use_addto);
+      }
+      return;
+    } catch (cudnn_frontend::cudnnException& e) {
```
catch中是否加一些提示信息比较好？

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,303 @@
+}
+
+template <typename T>
+void CudnnConvBwdFilterV8(DenseTensor* x_tensor,
```
`CudnnConvBwdDataV8`、`CudnnConvBwdFilterV8`这两个函数建议实现在conv_cudnn_frontend.h头文件中，因为后面还能复用到其他conv类算子中。

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel_impl_v8.h:
```diff
@@ -0,0 +1,303 @@
+}
+
+template <typename T, typename Context>
+void ConvCudnnGradKernelImplV8(DenseTensor* transformed_input,
```
这个函数在其他算子中应该不会被复用到吧，可以直接实现在conv_grad_kernel.cu中？

Xreki commented on paddle/phi/kernels/gpudnn/conv_grad_kernel.cu:
```diff
@@ -253,249 +552,61 @@ void ConvCudnnGradKernel(const Context& ctx,
+                                 groups,
+                                 &transformed_input_grad,
+                                 &transformed_filter_grad_channel);
+  else
```
这样写，代码理解起来，有点奇怪啊。。。

Xreki reviewed (commented):
Great work~

Xreki commented:
另外，需修复一下rocm平台上的编译错误：

![image](https://user-images.githubusercontent.com/12538138/201267029-6d46da7b-29b9-4d0b-9188-f27dda6d51b1.png)

Tom-Zheng committed:
Review fix:
- move functions in  conv_kernel_impl_v8.h and conv_grad_kernel_impl_v8.h to conv_kernel.cu and conv_grad_kernelk.cu
- add const specifier for input tensor
- add logging when plans fail to execute
- move CudnnConvBwdFilterV8 and CudnnConvBwdDataV8 to conv_cudnn_frontend.h

Tom-Zheng committed:
- move plan building outside of cache

Tom-Zheng committed:
test=cuda117

Tom-Zheng committed:
Fix ROCM build

Tom-Zheng committed:
test=cuda117

Tom-Zheng committed:
Fix ROCM build

Tom-Zheng committed:
test=cuda117

Xreki reviewed (approved):
LGTM. 相关功能和代码建议后续进一步优化，一方面和cudnn v7当前行为保持一致，另一方便后续便于将Frontend API应用到更多算子。

YuanRisheng commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,464 @@
+
+#include <vector>
+
+#include "paddle/fluid/framework/convert_utils.h"
```
最近PHI算子库在做解耦Fluid依赖的工作，convert_utils.h刚刚清理完毕，参考PR:https://github.com/PaddlePaddle/Paddle/pull/48001  这里建议参考该PR的方式把这个头文件移除

YuanRisheng commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,464 @@
+#include <vector>
+
+#include "paddle/fluid/framework/convert_utils.h"
+#include "paddle/fluid/platform/device/gpu/cuda/cudnn_desc.h"
```
这里还是不建议引入额外的Fluid头文件到PHi下，可以先在phi的backends/gpu/cuda目录下建立同样的头文件，把需要用到的函数拷贝过来，我看了这个文件里用到的应该不多，比较好处理

YuanRisheng commented on paddle/phi/kernels/gpudnn/conv_cudnn_frontend.h:
```diff
@@ -0,0 +1,464 @@
+        .setStrides(strides.size(), strides.data())
+        .setId(id)
+        .setAlignment(GetAlignment(tensor))
+        .setDataType(paddle::platform::ToCudnnDataType(
```
这里或许可以这样改：phi下新加一个ToCudnnDataType，逻辑和fluid下的一样，只不过传入的datatype类型不是proto::Vartype，而是phi的DataType

YuanRisheng commented:
该PR先合入，Fluid头文件让相关同学进行清理
